libyul/optimiser/UnusedStoreEliminator.cpp

```diff
@@ -0,0 +1,439 @@
+		values[name] = AssignedValue{expression, {}};
+	Expression const zeroLiteral{Literal{{}, LiteralKind::Number, YulString{"0"}, {}}};
+	Expression const oneLiteral{Literal{{}, LiteralKind::Number, YulString{"1"}, {}}};
+	Expression const thirtyTwoLiteral{Literal{{}, LiteralKind::Number, YulString{"32"}, {}}};
+	values[YulString{zero}] = AssignedValue{&zeroLiteral, {}};
+	values[YulString{one}] = AssignedValue{&oneLiteral, {}};
+	values[YulString{thirtyTwo}] = AssignedValue{&thirtyTwoLiteral, {}};
+
+	bool const ignoreMemory = MSizeFinder::containsMSize(_context.dialect, _ast);
+	UnusedStoreEliminator rse{
+		_context.dialect,
+		functionSideEffects,
+		ControlFlowSideEffectsCollector{_context.dialect, _ast}.functionSideEffectsNamed(),
+		values,
+		ignoreMemory
+	};
+	rse(_ast);
+	rse.changeUndecidedTo(State::Unused, Location::Memory);
+	rse.changeUndecidedTo(State::Used, Location::Storage);
+	rse.scheduleUnusedForDeletion();
+
+	StatementRemover remover(rse.m_pendingRemovals);
+	remover(_ast);
+}
+
+void UnusedStoreEliminator::operator()(FunctionCall const& _functionCall)
+{
+	UnusedStoreBase::operator()(_functionCall);
+
+	for (Operation const& op: operationsFromFunctionCall(_functionCall))
+		applyOperation(op);
+
+	ControlFlowSideEffects sideEffects;
+	if (auto builtin = m_dialect.builtin(_functionCall.functionName.name))
+		sideEffects = builtin->controlFlowSideEffects;
+	else
+		sideEffects = m_controlFlowSideEffects.at(_functionCall.functionName.name);
+
+	if (!sideEffects.canContinue)
+	{
+		changeUndecidedTo(State::Unused, Location::Memory);
+		changeUndecidedTo(sideEffects.canTerminate ? State::Used : State::Unused, Location::Storage);
+	}
+}
+
+void UnusedStoreEliminator::operator()(FunctionDefinition const& _functionDefinition)
+{
+	ScopedSaveAndRestore storeOperations(m_storeOperations, {});
+	UnusedStoreBase::operator()(_functionDefinition);
+}
+
+
+void UnusedStoreEliminator::operator()(Leave const&)
+{
+	changeUndecidedTo(State::Used);
+}
+
+void UnusedStoreEliminator::visit(Statement const& _statement)
+{
+	using evmasm::Instruction;
+
+	UnusedStoreBase::visit(_statement);
+
+	auto const* exprStatement = get_if<ExpressionStatement>(&_statement);
+	if (!exprStatement)
+		return;
+
+	FunctionCall const* funCall = get_if<FunctionCall>(&exprStatement->expression);
+	if (!funCall)
+		return;
+	optional<Instruction> instruction = toEVMInstruction(m_dialect, funCall->functionName.name);
+	if (!instruction)
+		return;
+
+	if (!ranges::all_of(funCall->arguments, [](Expression const& _expr) -> bool {
+		return get_if<Identifier>(&_expr) || get_if<Literal>(&_expr);
+	}))
+		return;
+
+	if (
+		*instruction == Instruction::SSTORE ||
+		(!m_ignoreMemory && (
+			*instruction == Instruction::EXTCODECOPY ||
+			*instruction == Instruction::CODECOPY ||
+			*instruction == Instruction::CALLDATACOPY ||
+			*instruction == Instruction::RETURNDATACOPY ||
+			*instruction == Instruction::MSTORE ||
+			*instruction == Instruction::MSTORE8
```

Comment:
Don't we have some abstract semantic property to cover this? Hm, in those "writing to memory" includes the call opcodes...

Reply:
You are right, even if we don't have such a property yet, we should create it. "No side effects apart from writing to memory".